File: scripts/buckets.py
"""Grip buckets (plan section 0), computed from word_events (lessons) + card_results (flashcards). One function, one truth;
the flashcards page carries the same rules in JS (docs/js/buckets.js) and this module is the reference the JS test compares to.

Display names (Medi, 2026-09-10): cold = Good; ice_cold = Mastered. Stored IDs, thresholds and history stay unchanged.

  ice_cold  5 consecutive successes across >= 3 days: first-try cards or independent lesson use (one per lesson date)
  cold      said unprompted in a lesson with no correction, or flashcard right first try
  shaky     right on second try, or said only after Amal said it (prompted)
  missed    Amal corrected it, Medi asked for it, or wrong twice in a row on cards
  new       ONLY a word explicitly marked new: Amal (after-link / chat confirmation) or Medi marked it for a lesson (amal_rules kind
            'new'), or it appeared in the Doc after a lesson but not in the snapshot before it (Doc diff; needs the live import).
            It stays new until practised: 5 first-try card rights on 2 different days. Nothing is inferred into 'new'
            (Medi 2026-09-05: "we don't have any more context for what words are new"). The chat/asked inference only
            proposes 'new?' candidates for Amal's after-link; it never buckets.
            New words never mix with missed words: they go through the strict review-and-repeat loop first
  never     no Medi signal yet. Amal never adds to usage, review dates or mastery.
Lesson evidence is ordered after cards on the same date, matching the existing date-only lesson precedence.
Missed words and words learned in the last 3 lessons get weight 3 (cards, sentence suggestions, homework)."""
import datetime, collections

RECENT_LESSONS = 3
NEW_DRILL_RIGHTS, NEW_DRILL_DAYS = 5, 2      # a new word leaves 'new' after 5 first-try rights on 2 different days (Medi 2026-09-05)


GRAMMAR_KINDS = ('article', 'gender', 'tense', 'plural')


def _signal_from_event(e):
    if e.get('speaker') != 'Medi' or e.get('prompted') is None:
        return None
    if e.get('correction') and e.get('miss_kind') in GRAMMAR_KINDS and not e.get('asked'):
        return 'shaky' if e.get('prompted') else 'cold'      # the word was known; the slip is grammar, counted separately
    if e.get('miss_kind') == 'choice':
        return 'shaky'                                        # known word, wrong place: shaky, not missed
    if e.get('correction') or e.get('asked'):
        return 'missed'
    if e.get('prompted'):
        return 'shaky'
    return 'cold'


SIGNAL_RANK = {'missed': 0, 'shaky': 1, 'cold': 2}


def _per_lesson_signals(evs):
    """One signal per lesson, not per event (Medi 2026-09-05, the lAzem case): a word Medi said on his own in a lesson is cold for
    that lesson even if he also echoed it after Amal later; any correction or 'asked' in the lesson makes it missed."""
    by_day = collections.OrderedDict()
    for e in evs:
        s = _signal_from_event(e)
        if s:
            by_day.setdefault(str(e['lesson_date']), []).append(s)
    out = []
    for d, sigs in by_day.items():
        if 'missed' in sigs:
            out.append((d, 'missed'))
        elif 'cold' in sigs:
            out.append((d, 'cold'))          # said unprompted at least once: a later echo does not downgrade it
        else:
            out.append((d, 'shaky'))
    return out


def _signal_from_cards(cards):
    """cards: chronological [{'ts','result','attempt'}] -> (bucket, streak, streak_days)."""
    if not cards:
        return None, 0, []
    streak, days = 0, []
    for c in reversed(cards):
        if c['result'] == 'got' and int(c.get('attempt') or 1) == 1:
            streak += 1
            d = str(c['ts'])[:10]
            if d not in days:
                days.append(d)
        else:
            break
    last = cards[-1]
    if streak >= 5 and len(days) >= 3:
        return 'ice_cold', streak, days
    if last['result'] == 'missed':
        if sum(1 for c in cards[-3:] if c['result'] == 'missed') >= 2:   # wrong twice (within the last three answers)
            return 'missed', 0, []
        return 'cold' if _was_ice(cards[:-1]) else 'shaky', 0, []      # one miss after an ice-cold streak -> cold
    if int(last.get('attempt') or 1) > 1:
        return 'shaky', streak, days
    return 'cold', streak, days


def _was_ice(cards):
    b, s, d = _signal_from_cards(cards) if cards else (None, 0, [])
    return b == 'ice_cold'


def _independent_use(e):
    """A detected Medi use with explicit no-help/no-correction evidence, not verified pronunciation."""
    return (e.get('speaker') == 'Medi' and e.get('prompted') is False
            and e.get('correction') is False and e.get('asked') is False
            and e.get('miss_kind') not in ('choice', 'unclear'))


def _known_help_or_correction(e):
    return (e.get('prompted') is True or e.get('correction') is True
            or e.get('asked') is True or e.get('miss_kind') == 'choice')


def progress_from_context(context):
    """Shared lesson/card mastery model. Card drill streak remains separate for New words."""
    cards = sorted(context['cards'], key=lambda c: (str(c['ts']), str(c.get('id') or '')))
    timeline = [(str(c['ts'])[:10], 0, str(c['ts']), i, 'card', c) for i, c in enumerate(cards)]
    timeline += [(d, 1, '', i, 'lesson', (signal, qualifies))
                 for i, (d, signal, qualifies) in enumerate(context['lessons'])]
    bucket, streak, days, last_cards = 'never', 0, [], []
    for day, _, _, _, source, value in sorted(timeline, key=lambda r: r[:4]):
        was_mastered = bucket == 'ice_cold'
        if source == 'card':
            last_cards.append(value)
            success = value['result'] == 'got' and int(value.get('attempt') or 1) == 1
            if value['result'] == 'missed':
                signal = 'cold' if was_mastered else ('missed' if sum(c['result'] == 'missed' for c in last_cards[-3:]) >= 2 else 'shaky')
            else:
                signal = 'cold' if success else 'shaky'
        else:
            signal, qualifies = value
            if qualifies is None:
                continue  # Unknown evidence neither earns mastery nor erases an established streak.
            success = signal == 'cold' and qualifies
        if success:
            streak += 1
            if day not in days:
                days.append(day)
        else:
            streak, days = 0, []
        bucket = 'ice_cold' if streak >= 5 and len(days) >= 3 else signal
    _, card_streak, card_days = _signal_from_cards(cards)
    signal = bucket
    if context['new'] and not (card_streak >= NEW_DRILL_RIGHTS and len(card_days) >= NEW_DRILL_DAYS):
        bucket = 'new'
    return {'bucket': bucket, 'lesson_signal': signal, 'streak': card_streak, 'streak_days': card_days,
            'mastery_streak': streak, 'mastery_days': days}


def compute(word_events, card_results, lesson_dates, today=None, confirmed_new=None, doc_before=None, introduced=None):
    """Returns {word_key: stats}. word_events rows need lesson_date, word_key, speaker, prompted, correction, asked, t_start.
    card_results rows need word_key, ts, result, attempt. lesson_dates = all lesson dates (ISO strings).
    confirmed_new = {(lesson_date, word_key)} marked new for that lesson by Amal or Medi (amal_rules kind 'new').
    doc_before = {lesson_date: {word_key}} words present in the Doc snapshot taken BEFORE that lesson; a word heard in a lesson
    and absent from that lesson's snapshot is new by the Doc rule; a date missing from the dict = unknown, rule skipped.
    introduced = {(lesson_date, word_key)} chat/asked inference; kept in the stats as 'new_candidate', never buckets."""
    confirmed_new = set(confirmed_new or ())
    doc_before = doc_before or {}
    introduced = set(introduced or ())
    marked_keys = {k for _, k in confirmed_new}
    lesson_dates = sorted(set(str(d) for d in lesson_dates))
    recent_dates = set(lesson_dates[-RECENT_LESSONS:])
    ev_by = collections.defaultdict(list)
    for e in word_events:
        ev_by[e['word_key']].append(e)
    cd_by = collections.defaultdict(list)
    for c in card_results:
        cd_by[c['word_key']].append(c)
    out = {}
    for key in set(ev_by) | set(cd_by):
        evs = sorted(ev_by[key], key=lambda e: (str(e['lesson_date']), e.get('t_start') or 0))
        cards = sorted(cd_by[key], key=lambda c: str(c['ts']))
        medi = [e for e in evs if e.get('speaker') == 'Medi']
        independent = [e for e in medi if _independent_use(e)]
        qualifying_dates = {str(e['lesson_date']) for e in independent}
        helped_dates = {str(e['lesson_date']) for e in medi if _known_help_or_correction(e)}
        lesson_signals = _per_lesson_signals(medi)
        first_lesson = str(medi[0]['lesson_date']) if medi else None
        seen_dates = sorted({str(e['lesson_date']) for e in medi})
        # Introduction semantics are independent of who spoke; don't lose existing New marks.
        all_dates = sorted({str(e['lesson_date']) for e in evs})
        marked = key in marked_keys
        by_doc = any(d in doc_before and key not in doc_before[d] for d in all_dates)
        new_candidate = bool(first_lesson) and ((first_lesson, key) in introduced
                                                or any(e.get('asked') for e in evs if str(e['lesson_date']) == first_lesson))
        context = {'version': 2, 'new': marked or by_doc,
                   'lessons': [[d, sig, True if d in qualifying_dates else (False if d in helped_dates else None)] for d, sig in lesson_signals],
                   'cards': [{k: c[k] for k in ('id', 'ts', 'result', 'attempt') if k in c} for c in cards]}
        progress = progress_from_context(context)
        bucket = progress['bucket']
        last_reviewed = max([x for x in [seen_dates[-1] if seen_dates else None, str(cards[-1]['ts']) if cards else None] if x], default=None)
        recent = first_lesson in recent_dates if first_lesson else False
        out[key] = {'word_key': key, 'bucket': bucket, 'last_reviewed': last_reviewed, 'last_lesson': seen_dates[-1] if seen_dates else None,
                    'seen_lessons': len(seen_dates), 'times_seen': len(medi), 'independent_uses': len(independent),
                    'times_missed': sum(1 for e in medi if _signal_from_event(e) == 'missed') + sum(1 for c in cards if c['result'] == 'missed'),
                    'card_right': sum(1 for c in cards if c['result'] == 'got'), 'card_wrong': sum(1 for c in cards if c['result'] == 'missed'),
                    **progress, 'progress_context': context, 'recent': recent, 'weight': 3.0 if bucket in ('missed', 'new') else 1.0,
                    'new_candidate': new_candidate, 'grammar_misses': sum(1 for e in medi if e.get('correction') and e.get('miss_kind') in GRAMMAR_KINDS),
                    'grammar_kinds': sorted({e['miss_kind'] for e in medi if e.get('correction') and e.get('miss_kind') in GRAMMAR_KINDS})}
    return out


def introduced_from_chat(typed_rows, words=None):
    """{(lesson_date, word_key)} for every Doc word Amal typed in a lesson chat, by exact/loose/skeleton match of the typed form
    or by consonant family (Basa6tek -> the babse6 family), so a conjugation she typed counts for its Doc lemma."""
    import re
    from arabizi import Matcher
    import understand_lesson as ul
    words = words or ul.load_words()
    m = Matcher(words)
    fam = {}
    for w in words:
        fk = ul.doc_family(w)
        if fk:
            fam.setdefault(fk, set()).add(w['key'])
    out = set()
    for r in typed_rows:
        d = str(r['lesson_date'])
        for form in re.split(r'\s*/\s*|\n', str(r.get('text') or '')):
            form = form.strip(' ?.,!')
            if not form:
                continue
            k = m.match(form, fuzzy=False)
            if k:
                out.add((d, k))
            sk = ul.arabizi_to_ar_skel(form)                     # full consonant skeleton of the typed form (Basa6tek -> بستتك)
            for fk, keys in fam.items():                          # a Doc family is inside it with at most 4 affix consonants
                if len(fk) >= 3 and fk in sk and len(sk) - len(fk) <= 4:
                    out.update((d, k2) for k2 in keys)
    return out


def recompute_and_store():
    """Reads word_events + card_results + lessons from Supabase, writes word_stats. Returns the stats dict."""
    import db
    evs = db.select('word_events', {'select': 'lesson_date,word_key,speaker,prompted,correction,asked,miss_kind,t_start'})
    cards = db.select('card_results', {'select': 'id,word_key,ts,result,attempt'})
    dates = [r['date'] for r in db.select('lessons', {'select': 'date'})]
    typed = db.select('lesson_events', {'select': 'lesson_date,text', 'kind': 'eq.typed'})
    marks = db.select('amal_rules', {'select': 'lesson_date,word_key,kind', 'kind': 'eq.new'})
    confirmed = {(str(r['lesson_date']), r['word_key']) for r in marks if r.get('word_key')}
    stats = compute(evs, cards, dates, confirmed_new=confirmed, introduced=introduced_from_chat(typed))
    now = datetime.datetime.now(datetime.timezone.utc).isoformat()
    rows = []
    for s in stats.values():
        r = dict(s); r.pop('new_candidate', None); r['updated_at'] = now
        if r['last_reviewed'] and len(r['last_reviewed']) == 10:
            r['last_reviewed'] = r['last_reviewed'] + 'T00:00:00Z'
        rows.append(r)
    db.upsert('word_stats', rows, on='word_key')
    # rows for words that no longer have any event or card are stale (earlier builds, purged test rows): remove them
    live = set(stats)
    for r in db.select('word_stats', {'select': 'word_key'}):
        if r['word_key'] not in live:
            db.rest('DELETE', 'word_stats', params={'word_key': f"eq.{r['word_key']}"}, prefer='return=minimal')
    return stats

File: docs/js/buckets.js
// Grip buckets — JS port of scripts/buckets.py (the Python file is the reference; tests/test_m5_cards.py checks parity).
// Speaking and Flashcards are isolated. Mastered: 5 successes in that lane on >=3 dates.
// Missed recovery: two consecutive independent successes, first -> Shaky, second -> Good.
// shaky: prompted / right on second try. missed: corrected / wrong twice in a row on cards. never: no Medi signal.
(function (root) {
  // Display names only: keep stored IDs and scoring rules compatible with saved progress.
  const DISPLAY_LABELS = Object.freeze({ cold: 'Good', ice_cold: 'Mastered' });
  function label(bucket) { return Object.prototype.hasOwnProperty.call(DISPLAY_LABELS, bucket) ? DISPLAY_LABELS[bucket] : String(bucket ?? '').replace(/_/g, ' '); }
  const RECENT_LESSONS = 3, NEW_DRILL_RIGHTS = 5, NEW_DRILL_DAYS = 2;
  const GRAMMAR_KINDS = ['article', 'gender', 'tense', 'plural'];
  function signalFromEvent(e) {
    if (e.speaker !== 'Medi' || e.prompted === null || e.prompted === undefined) return null;
    if (e.correction && GRAMMAR_KINDS.includes(e.miss_kind) && !e.asked) return e.prompted ? 'shaky' : 'cold';
    if (e.miss_kind === 'choice') return 'shaky';
    if (e.correction || e.asked) return 'missed';
    if (e.prompted) return 'shaky';
    return 'cold';
  }
  function wasIce(cards) { return cards.length ? signalFromCards(cards)[0] === 'ice_cold' : false; }
  function signalFromCards(cards) {
    if (!cards.length) return [null, 0, []];
    let streak = 0; const days = [];
    for (let i = cards.length - 1; i >= 0; i--) {
      const c = cards[i];
      if (c.result === 'got' && (parseInt(c.attempt || 1) === 1)) { streak++; const d = String(c.ts).slice(0, 10); if (!days.includes(d)) days.push(d); }
      else break;
    }
    const last = cards[cards.length - 1];
    if (streak >= 5 && days.length >= 3) return ['ice_cold', streak, days];
    if (last.result === 'missed') {
      if (cards.slice(-3).filter(c => c.result === 'missed').length >= 2) return ['missed', 0, []];   // wrong twice within the last three
      return [wasIce(cards.slice(0, -1)) ? 'cold' : 'shaky', 0, []];
    }
    if (parseInt(last.attempt || 1) > 1) return ['shaky', streak, days];
    return ['cold', streak, days];
  }
  function independentUse(e) {
    return e.speaker === 'Medi' && e.prompted === false && e.correction === false && e.asked === false && !['choice','unclear'].includes(e.miss_kind);
  }
  function isSpoken(e) { return e.speaker==='Medi' && typeof e.t_start==='number' && Number.isFinite(e.t_start) && e.t_start>=0 && !String(e.text||'').toLowerCase().startsWith('homework:'); }
  function trackProgress(context,lane) {
    const cards=(lane==='flashcards'?context.cards:[]).slice().sort((a,b)=>String(a.ts).localeCompare(String(b.ts)) || String(a.id||'').localeCompare(String(b.id||'')));
    const timeline=cards.map((c,i)=>({day:String(c.ts).slice(0,10),rank:0,ts:String(c.ts),i,source:'card',value:c}));
    for(const [i,[day,signal,qualifies]] of (lane==='speaking'?context.lessons:[]).entries()) timeline.push({day,rank:1,ts:'',i,source:'lesson',value:[signal,qualifies]});
    timeline.sort((a,b)=>a.day.localeCompare(b.day)||a.rank-b.rank||a.ts.localeCompare(b.ts)||a.i-b.i);
    let bucket='never', masteryStreak=0, masteryDays=[], recoveryLeft=0; const lastCards=[];
    for(const event of timeline) {
      const wasMastered=bucket==='ice_cold'; let signal, success;
      if(event.source==='card') {
        const c=event.value; lastCards.push(c);
        success=c.result==='got' && parseInt(c.attempt||1)===1;
        signal=c.result==='missed' ? (wasMastered?'cold':(lastCards.slice(-3).filter(c=>c.result==='missed').length>=2?'missed':'shaky')) : (success?'cold':'shaky');
      } else { const [s,qualifies]=event.value; if(qualifies===null) continue; signal=s; success=s==='cold' && qualifies; }
      if(success) {
        masteryStreak++; if(!masteryDays.includes(event.day)) masteryDays.push(event.day);
        if(recoveryLeft) { recoveryLeft--; signal=recoveryLeft?'shaky':'cold'; }
      } else {
        masteryStreak=0; masteryDays=[];
        if(signal==='missed'||recoveryLeft) { recoveryLeft=2; if(signal==='cold') signal='shaky'; }
      }
      bucket=masteryStreak>=5 && masteryDays.length>=3?'ice_cold':signal;
    }
    return {bucket,streak:masteryStreak,streak_days:masteryDays,mastery_streak:masteryStreak,mastery_days:masteryDays,recovery_left:recoveryLeft};
  }
  function progressFromContext(context) {
    if(context.version!==3) throw new Error('Speaking provenance must be rebuilt before scoring old mixed contexts');
    const speaking={...trackProgress(context,'speaking'),...context.speaking}, rawSpeaking=speaking.bucket;
    if(context.new && speaking.intro_uses<5) speaking.bucket='new';
    speaking.intro_target=context.new?5:0; speaking.weight=['new','missed'].includes(speaking.bucket)?3:1;
    const cards=context.cards.slice().sort((a,b)=>String(a.ts).localeCompare(String(b.ts))||String(a.id||'').localeCompare(String(b.id||'')));
    const flashcards=trackProgress(context,'flashcards');
    if(context.new && !(flashcards.streak>=NEW_DRILL_RIGHTS && flashcards.streak_days.length>=NEW_DRILL_DAYS)) flashcards.bucket='new';
    Object.assign(flashcards,{last_reviewed:cards.length?String(cards[cards.length-1].ts):null,attempts:cards.length,
      card_right:cards.filter(c=>c.result==='got').length,times_missed:cards.filter(c=>c.result==='missed').length,
      first_try_right:cards.filter(c=>c.result==='got'&&parseInt(c.attempt||1)===1).length});
    flashcards.weight=['new','missed'].includes(flashcards.bucket)?3:1;
    return {...Object.fromEntries(['bucket','streak','streak_days','mastery_streak','mastery_days','last_reviewed','times_missed','weight'].map(k=>[k,speaking[k]])),
      lesson_signal:rawSpeaking,card_right:flashcards.card_right,card_wrong:flashcards.times_missed,
      progress_scores:{version:1,speaking,flashcards}};
  }
  function mergeProgress(stats, log) {
    // Merge durable server card records and unsynced local IDs, not activity dates used as sync cursors.
    const source=stats.progress_context, cards=source.cards.map(c=>({...c}));
    const ids=new Set(cards.filter(c=>c.id).map(c=>c.id));
    for(const c of log) {
      if(c.id ? ids.has(c.id) : cards.some(old=>!old.id && old.ts===c.ts && old.result===c.result && (old.attempt||1)===(c.attempt||1))) continue;
      cards.push({...c}); if(c.id) ids.add(c.id);
    }
    cards.sort((a,b)=>String(a.ts).localeCompare(String(b.ts))||String(a.id||'').localeCompare(String(b.id||'')));
    const context={...source,cards};
    return {...stats,...progressFromContext(context),progress_context:context};
  }
  function emptyStats(key) {
    const context={version:3,new:false,lessons:[],cards:[],speaking:{times_seen:0,independent_uses:0,times_missed:0,last_reviewed:null,seen_lessons:0,intro_uses:0,new_since:null}};
    return {word_key:key,bucket:'never',recent:false,times_seen:0,independent_uses:0,times_missed:0,seen_lessons:0,
      last_lesson:null,last_reviewed:null,card_right:0,card_wrong:0,
      ...progressFromContext(context),progress_context:context};
  }
  function mergeStats(stats, log) {
    const out=Object.fromEntries(Object.entries(stats).filter(([,s])=>s.progress_context?.version===3).map(([key,s])=>[key,mergeProgress(s,[])])), by=new Map();
    for(const row of log||[]) {
      if(!row.word_key || !['got','missed'].includes(row.result) || !row.ts) continue;
      if(!by.has(row.word_key)) by.set(row.word_key,[]); by.get(row.word_key).push(row);
    }
    for(const [key,rows] of by) {
      const s=out[key]||emptyStats(key);
      if(s.progress_context?.version!==3) continue; // Never reinterpret legacy mixed scores as Speaking.
      const merged=mergeProgress(s,rows);
      out[key]={...merged,weight:['missed','new'].includes(merged.bucket)?3:1};
    }
    return out;
  }
  function compute(wordEvents, cardResults, lessonDates, confirmedNew, docBefore) {
    // confirmedNew: Set of `${lesson_date}|${word_key}` marked new by Amal/Medi; docBefore: {lesson_date: Set(word_key)} in the Doc before that lesson
    confirmedNew = confirmedNew || new Set(); docBefore = docBefore || {};
    const markedKeys = new Set([...confirmedNew].map(x => x.split('|').slice(1).join('|')));
    const dates = [...new Set(lessonDates.map(String))].sort();
    const recent = new Set(dates.slice(-RECENT_LESSONS));
    const evBy = {}, cdBy = {};
    for (const e of wordEvents) (evBy[e.word_key] = evBy[e.word_key] || []).push(e);
    for (const c of cardResults) (cdBy[c.word_key] = cdBy[c.word_key] || []).push(c);
    const out = {};
    for (const key of new Set([...Object.keys(evBy), ...Object.keys(cdBy), ...markedKeys])) {
      const evs = (evBy[key] || []).slice().sort((a, b) => String(a.lesson_date).localeCompare(String(b.lesson_date)) || ((a.t_start || 0) - (b.t_start || 0)));
      const cards = (cdBy[key] || []).slice().sort((a, b) => String(a.ts).localeCompare(String(b.ts)));
      const medi=evs.filter(isSpoken), independent=medi.filter(independentUse);
      const qualifyingDates=new Set(independent.map(e=>String(e.lesson_date)));
      const helpedDates=new Set(medi.filter(e=>e.prompted===true||e.correction===true||e.asked===true||e.miss_kind==='choice').map(e=>String(e.lesson_date)));
      const byDay = new Map();
      for (const e of medi) { const s = signalFromEvent(e); if (s) { const d = String(e.lesson_date); if (!byDay.has(d)) byDay.set(d, []); byDay.get(d).push(s); } }
      const lessonSignals = [...byDay.entries()].map(([d, sigs]) => [d, sigs.includes('missed') ? 'missed' : (sigs.includes('cold') ? 'cold' : 'shaky')]);   // one signal per lesson: unprompted use beats a later echo
      const firstLesson = medi.length ? String(medi[0].lesson_date) : null;
      const seenDates = [...new Set(medi.map(e => String(e.lesson_date)))].sort();
      const marked = markedKeys.has(key);
      const byDoc = evs.some(e => docBefore[String(e.lesson_date)] && !docBefore[String(e.lesson_date)].has(key));
      const newDates=[...[...confirmedNew].filter(x=>x.split('|').slice(1).join('|')===key).map(x=>x.split('|')[0]),
        ...evs.filter(e=>docBefore[String(e.lesson_date)]&&!docBefore[String(e.lesson_date)].has(key)).map(e=>String(e.lesson_date))].sort();
      const newSince=newDates[0]||null;
      const speaking={times_seen:medi.length,independent_uses:independent.length,times_missed:medi.filter(e=>signalFromEvent(e)==='missed').length,
        last_reviewed:seenDates[seenDates.length-1]||null,seen_lessons:seenDates.length,intro_uses:medi.filter(e=>!newSince||String(e.lesson_date)>=newSince).length,new_since:newSince};
      const context={version:3,new:marked||byDoc,speaking,lessons:lessonSignals.map(([d,s])=>[d,s,qualifyingDates.has(d)?true:(helpedDates.has(d)?false:null)]),
        cards:cards.map(c=>Object.fromEntries(['id','ts','result','attempt'].filter(k=>k in c).map(k=>[k,c[k]])))};
      const progress=progressFromContext(context), bucket=progress.bucket;
      const lastReviewed = seenDates[seenDates.length - 1] || null;
      const isRecent = firstLesson ? recent.has(firstLesson) : false;
      out[key] = { word_key: key, bucket, last_reviewed: lastReviewed, last_lesson:seenDates[seenDates.length-1]||null, seen_lessons: seenDates.length, times_seen: medi.length, independent_uses:independent.length,
        times_missed: medi.filter(e => signalFromEvent(e)==='missed').length,
        card_right: cards.filter(c => c.result === 'got').length, card_wrong: cards.filter(c => c.result === 'missed').length,
        ...progress,progress_context:context,recent: isRecent, weight: (bucket === 'missed' || bucket === 'new') ? 3 : 1,
        grammar_misses:medi.filter(e=>e.correction&&GRAMMAR_KINDS.includes(e.miss_kind)).length,
        grammar_kinds:[...new Set(medi.filter(e=>e.correction&&GRAMMAR_KINDS.includes(e.miss_kind)).map(e=>e.miss_kind))].sort() };
    }
    return out;
  }
  root.AneesBuckets = { compute, signalFromCards, signalFromEvent, label, independentUse, progressFromContext, mergeProgress, emptyStats, mergeStats };
})(typeof window !== 'undefined' ? window : globalThis);
